src/lib/components/Identifiers/PIDField.js

```diff
@@ -0,0 +1,506 @@
+};
+
+/**
+ * Render managed or unamanged PID fields and update
+ * Formik form on input changed.
+ * The field value has the following format:
+ * { 'doi': { identifier: '<value>', provider: '<value>', client: '<value>' } }
+ */
+class CustomPIDField extends Component {
+  constructor(props) {
+    super(props);
+
+    const { canBeManaged, canBeUnmanaged, field } = this.props;
+    this.canBeManagedAndUnmanaged = canBeManaged && canBeUnmanaged;
+
+    // init state
+    const canBeOnlyManaged = canBeManaged && !canBeUnmanaged;
+
+    const currentIdentifier = _get(field.value, 'identifier', '');
+    let managedIdentifier = '',
+      unmanagedIdentifier = '';
+    if (currentIdentifier !== '') {
+      const currentProvider = _get(field.value, 'provider', '');
+      const isProviderExternal = currentProvider === PROVIDER_EXTERNAL;
+      managedIdentifier = !isProviderExternal ? currentIdentifier : '';
+      unmanagedIdentifier = isProviderExternal ? currentIdentifier : '';
+    }
+
+    this.state = {
+      isManagedSelected: canBeOnlyManaged || false,
+      managedIdentifier: managedIdentifier,
+      unmanagedIdentifier: unmanagedIdentifier,
+    };
+  }
+
+  updateIdentifer = (form, fieldPath, pid) => {
+    this.debounced && this.debounced.cancel();
+    this.debounced = _debounce(() => {
+      //PIDS-FIXME
+      console.log('TODO implement onIdentifierChange', pid);
+      form.setFieldValue(fieldPath, pid);
+    }, UPDATE_PID_DEBOUNCE_MS);
+    this.debounced();
+  };
+
+  buildUnmanagedPid = (identifier) => {
+    return { identifier: identifier, provider: PROVIDER_EXTERNAL };
+  };
+
+  buildManagedPid = (identifier, provider) => {
+    const { pidClient } = this.props;
+    const pidValue = {
+      identifier: identifier,
+    };
+
+    if (identifier) {
+      pidValue.provider = provider;
+      pidValue.client = pidClient;
+    }
+    return pidValue;
+  };
+
+  buildPid = (isManagedSelected, managedIdentifier, unmanagedIdentifier) => {
+    const { pidProvider } = this.props;
+    return isManagedSelected
+      ? this.buildManagedPid(managedIdentifier, pidProvider)
+      : this.buildUnmanagedPid(unmanagedIdentifier);
+  };
+
+  render() {
+    const {
+      isManagedSelected,
+      managedIdentifier,
+      unmanagedIdentifier,
+    } = this.state;
+    const {
+      btnLabelGetPID,
+      form,
+      fieldPath,
+      managedHelpText,
+      pidLabel,
+      pidIcon,
+      pidPlaceholder,
+      required,
+      unmanagedHelpText,
+    } = this.props;
+    return (
+      <>
+        <Form.Field required={required}>
+          <FieldLabel htmlFor={fieldPath} icon={pidIcon} label={pidLabel} />
+        </Form.Field>
+
+        {this.canBeManagedAndUnmanaged && (
+          <ManagedUnmanagedSwitch
+            onManagedUnmanagedChange={(isManagedSelected) => {
+              this.setState({
+                isManagedSelected: isManagedSelected,
+              });
+              const pid = this.buildPid(
```

Comment:
Are you assuming here that the `pid` variable will be set after the state has changed above? If so, then you will need to pass a callback function to the `setState()` above :)

Reply:
thanks for noticing, but this is on purpose, no need to have the state updated ;)